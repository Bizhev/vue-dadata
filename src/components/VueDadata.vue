<template>
  <div class="vue-dadata vue-dadata__container">
    <div class="vue-dadata__search">
      <input
        type="text"
        class="vue-dadata__input"
        :disabled="disabled"
        :placeholder="placeholder"
        v-model="inputQuery"
        :autoComplete="autocomplete"
        @change="onInputChange"
        @keydown="onKeyPress"
        @focus="onInputFocus"
        @blur="onInputBlur"
      />
    </div>
    <div class="vue-dadata__suggestions"></div>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator';
import DadataAddress, { BoundsType } from '@/types/DadataAddress';
import DadataSuggestion from '@/types/DadataSuggestion';
import axios from 'axios';
import getSuggestions from '@/api/getSuggestions';

@Component({
  name: 'VueDadata',
})
export default class VueDadata extends Vue {
  @Prop({ type: String, required: true }) public readonly token!: string;
  @Prop(String) public readonly query?: string;
  @Prop(String) public readonly placeholder?: string;
  @Prop(Boolean) public readonly autoload?: boolean;
  @Prop(Number) public readonly count?: number;
  @Prop(String) public readonly autocomplete?: string;
  @Prop(Boolean) public readonly disabled?: boolean;
  @Prop(String) public readonly fromBound?: BoundsType;
  @Prop(String) public readonly toBound?: BoundsType;
  @Prop(String) public readonly address?: DadataSuggestion;
  @Prop(Function) public readonly onChange?: (
    suggestion: DadataSuggestion[],
  ) => void;
  @Prop(Function) public readonly validate?: (value: string) => void;

  public inputQuery: string = '';
  public inputFocused: boolean = false;
  public suggestions: DadataSuggestion[] = [];
  public suggestionIndex: number = -1;
  public suggestionsVisible: boolean = true;
  public isValid: boolean = false;

  public created() {
    // console.log('created');
    // TODO
  }

  public onInputFocus() {
    // console.log('onInputFocus');
    // TODO
  }

  public onInputBlur() {
    // console.log('onInputBlur');
    // TODO
  }

  public onInputChange() {
    // console.log('onInputChange');
    // TODO
  }

  public onKeyPress(event: KeyboardEvent) {
    // console.log('onKeyPress');
    // TODO
    if (event.which === 13) {
      // enter
      this.fetchSuggestions();
    }
  }

  public onSuggestionClick() {
    // console.log('onSuggestionClick');
    // TODO
  }

  private async fetchSuggestions(): Promise<DadataSuggestion[]> {
    const suggestions = await getSuggestions({
      token: this.token,
      query: this.inputQuery,
    });

    return suggestions;
  }
}
</script>

<style lang="scss">
.vue-dadata {
}
</style>
