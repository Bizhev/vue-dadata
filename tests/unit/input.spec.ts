import { mount } from '@vue/test-utils';
import VueDadata from '@/components/VueDadata.vue';

describe('VueDadata.vue', () => {
  it('Testing working token and query', async () => {
    const wrapper = mount(VueDadata, {
        propsData: {
            token: process.env.VUE_APP_DADATA_API_KEY
        }
    });

    wrapper.find('input').setValue('Москва')
    wrapper.setData({inputFocused: true, suggestionsVisible: true})
    await new Promise((r) => setTimeout(r, 2000));
    expect(wrapper.find('.vue-dadata__suggestions-item').exists()).toBe(true)
    expect(wrapper.element).toMatchSnapshot()
  });
});