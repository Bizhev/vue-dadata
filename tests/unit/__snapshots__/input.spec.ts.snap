// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VueDadata.vue Testing working token and query 1`] = `
<div
  class="vue-dadata vue-dadata__container"
>
  <div
    class="vue-dadata__search"
  >
    <input
      class="vue-dadata__input"
      type="text"
    />
  </div>
   
  <div
    class="vue-dadata__suggestions"
    id="suggestions"
  >
    <span
      class="vue-dadata__suggestions-item"
    >
      <span
        class=""
      >
        г 
      </span>
      <mark
        class="vue-dadata__suggestions-item-highlight "
      >
        Москва
      </mark>
    </span>
    <span
      class="vue-dadata__suggestions-item"
    >
      <span
        class=""
      >
        Нижегородская обл, Лукояновский р-н, поселок Новая 
      </span>
      <mark
        class="vue-dadata__suggestions-item-highlight "
      >
        Москва
      </mark>
    </span>
    <span
      class="vue-dadata__suggestions-item"
    >
      <span
        class=""
      >
        Приморский край, Шкотовский р-н, деревня Новая 
      </span>
      <mark
        class="vue-dadata__suggestions-item-highlight "
      >
        Москва
      </mark>
    </span>
    <span
      class="vue-dadata__suggestions-item"
    >
      <span
        class=""
      >
        респ Башкортостан, Бижбулякский р-н, село Менеуз-
      </span>
      <mark
        class="vue-dadata__suggestions-item-highlight "
      >
        Москва
      </mark>
    </span>
    <span
      class="vue-dadata__suggestions-item"
    >
      <span
        class=""
      >
        Псковская обл, Порховский р-н, деревня 
      </span>
      <mark
        class="vue-dadata__suggestions-item-highlight "
      >
        Москва
      </mark>
    </span>
    <span
      class="vue-dadata__suggestions-item"
    >
      <span
        class=""
      >
        Брянская обл, Красногорский р-н, поселок Новая 
      </span>
      <mark
        class="vue-dadata__suggestions-item-highlight "
      >
        Москва
      </mark>
    </span>
    <span
      class="vue-dadata__suggestions-item"
    >
      <span
        class=""
      >
        Кировская обл, Верхошижемский р-н, деревня 
      </span>
      <mark
        class="vue-dadata__suggestions-item-highlight "
      >
        Москва
      </mark>
    </span>
    <span
      class="vue-dadata__suggestions-item"
    >
      <span
        class=""
      >
        Тамбовская обл, рп Мучкапский, ул Красная 
      </span>
      <mark
        class="vue-dadata__suggestions-item-highlight "
      >
        Москва
      </mark>
    </span>
    <span
      class="vue-dadata__suggestions-item"
    >
      <span
        class=""
      >
        Тверская обл, Пеновский р-н, деревня 
      </span>
      <mark
        class="vue-dadata__suggestions-item-highlight "
      >
        Москва
      </mark>
    </span>
    <span
      class="vue-dadata__suggestions-item"
    >
      <span
        class=""
      >
        Московская обл, Ногинский р-н, тер. снт Вечерняя 
      </span>
      <mark
        class="vue-dadata__suggestions-item-highlight "
      >
        Москва
      </mark>
    </span>
  </div>
</div>
`;
